### PR TITLE
Add ProGuard/R8 rules

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

**Review by Egorand — APPROVED** [review]

File: zipline/src/jvmMain/resources/META-INF/proguard/zipline.pro [hunk 1]
```diff
@@ -0,0 +1,36 @@
+### ZIPLINE
+
+-keepclassmembers class app.cash.zipline.QuickJsException { * ; }
```
Comment (JakeWharton) [inline zipline/src/jvmMain/resources/META-INF/proguard/zipline.pro]: This is insufficient as the type is looked up by FQCN: https://github.com/cashapp/zipline/blob/d7ae3d7bde4e1beee119a0d460cd22cc223f3b77/zipline/native/Context.cpp#L94-L95
```suggestion
-keep,allowoptimization class app.cash.zipline.QuickJsException { * ; }
```

File: zipline/src/jvmMain/resources/META-INF/proguard/zipline.pro [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+-keepclassmembers interface app.cash.zipline.InterruptHandler { * ; }
```
Comment (JakeWharton) [inline zipline/src/jvmMain/resources/META-INF/proguard/zipline.pro]: This is insufficient as the type is looked up by FQCN: https://github.com/cashapp/zipline/blob/d7ae3d7bde4e1beee119a0d460cd22cc223f3b77/zipline/native/Context.cpp#L103
```suggestion
-keep,allowoptimization interface app.cash.zipline.InterruptHandler { * ; }
```

File: zipline/src/jvmMain/resources/META-INF/proguard/zipline.pro [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+-keepclassmembers interface app.cash.zipline.internal.bridge.CallChannel { * ; }
```
Comment (JakeWharton) [inline zipline/src/jvmMain/resources/META-INF/proguard/zipline.pro]: This is insufficient as the type is looked up by FQCN: https://github.com/cashapp/zipline/blob/trunk/zipline/native/OutboundCallChannel.cpp#L26
```suggestion
-keep,allowoptimization interface app.cash.zipline.internal.bridge.CallChannel { * ; }
```

File: zipline/src/jvmMain/resources/META-INF/proguard/zipline.pro [hunk 4]
```diff
@@ -0,0 +1,36 @@
+### ZIPLINE
```
Comment (JakeWharton) [inline zipline/src/jvmMain/resources/META-INF/proguard/zipline.pro]: This file needs set as `consumerProguardFiles` for the Android build to also use them.

**Review by JakeWharton — APPROVED** [review]

File: zipline/src/jvmMain/resources/META-INF/proguard/zipline.pro (hunk 4 above)
Comment (swankjesse) [reply zipline/src/jvmMain/resources/META-INF/proguard/zipline.pro]: Oooh fixed. I used Moshi as a sample but it’s a JAR and not an AAR. Should have tried something else.

Comment (JakeWharton) [reply zipline/src/jvmMain/resources/META-INF/proguard/zipline.pro]: should just be something like `consumerProguardFiles files("src/jvmMain/resources/META-INF/proguard/zipline.pro")` inside... some block.

Comment (JakeWharton) [reply zipline/src/jvmMain/resources/META-INF/proguard/zipline.pro]: `android.defaultConfig`: https://github.com/JakeWharton/butterknife/blob/fcdebedf3276096db2f51bf6372b849b5a9c75ed/butterknife/build.gradle#L3-L9

**Comment by JakeWharton** [comment]
It's still useful to bundle these in the JVM artifact as a JVM consumer could be using R8 to shrink their application.

**Pull request merged**